src/node_id.rs

```diff
@@ -0,0 +1,334 @@
+        self.discohash.cmp(&other.discohash)
+    }
+}
+
+impl PartialOrd for NodeId {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl From<PublicKey> for NodeId {
+    #[inline]
+    fn from(pubkey: PublicKey) -> NodeId {
+        let keyhash = hash(pubkey.as_bytes(), 32);
+        NodeId { discohash: keyhash }
+    }
+}
+
+impl PartialEq for NodeId {
+    fn eq(&self, other: &NodeId) -> bool {
+        self.discohash == other.discohash
+    }
+}
+
+impl PartialEq<NodeInfo> for NodeId {
+    fn eq(&self, other: &NodeInfo) -> bool {
+        *self == other.id
+    }
+}
+
+impl PartialEq<Vec<u8>> for NodeId {
+    fn eq(&self, other: &Vec<u8>) -> bool {
+        self.discohash == *other
+    }
+}
+
+pub trait KadMetric: PartialEq + Clone + fmt::Debug {
+    type Err;
+    type Metric;
+
+    fn distance(&self, other: &Self) -> Result<Self, Self::Err>;
+    // used in `store` specifically 
+    fn metric_distance(&self, other: &Self) -> Result<Self::Metric, Self::Err>;
+}
+
+impl KadMetric for NodeId {
+    type Err = DistanceIsZero;
+    type Metric = U256;
+
+    fn distance(&self, other: &NodeId) -> Result<NodeId, DistanceIsZero> {
+        let dist = self
+            .discohash
+            .iter()
+            .zip(other.discohash.iter())
+            .map(|(first, second)| first ^ second)
+            .collect();
+        let metric_node = NodeId { discohash: dist };
+        if metric_node.is_zero() {
+            return Err(DistanceIsZero);
+        } else {
+            return Ok(metric_node);
+        }
+    }
+
+    fn metric_distance(&self, other: &NodeId) -> Result<U256, DistanceIsZero> {
+        let a = U256::from(self.discohash.clone().as_slice());
+        let b = U256::from(other.discohash.clone().as_slice());
+        // xor
+        let distance = a ^ b;
+        if distance == U256::from(0) {
+            return Err(DistanceIsZero);
+        } else {
+            return Ok(distance)
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::{KadMetric, NodeId};
+    use crate::ed25519::Keypair;
+    use crate::error::{ParseError, DistanceIsZero};
+    use disco::hash;
+    use rand;
+
+    // TODO: change when error becomes more descriptive
+    // for the moment, all ParseError == ParseError
+    impl PartialEq for ParseError {
+        fn eq(&self, other: &ParseError) -> bool {
+            true
+        }
+    }
+
+    // \A DistanceIsZero == DistanceIsZero
+    impl PartialEq for DistanceIsZero {
+        fn eq(&self, other: &DistanceIsZero) -> bool {
+            true
+        }
+    }
+
+    #[test]
+    fn node_id_is_public_key() {
+        let key = Keypair::generate(&mut rand::thread_rng());
+        let keyhash = hash(key.public.as_bytes(), 32);
+        let node_id = NodeId { discohash: keyhash };
+        assert!(node_id.is_public_key(key.public));
+    }
+
+    #[test]
+    fn random_node_id_is_valid() {
```

Comment:
what is the purpose of this test? Are you checking that the hash function works correctly?

Reply:
true, this is useless and I moved the `random` to `util::test`